driver/k8s/logicalvolume_service.go

```diff
@@ -108,25 +109,49 @@ func NewLogicalVolumeService(mgr manager.Manager) (*LogicalVolumeService, error)
 }
 
 // CreateVolume creates volume
-func (s *LogicalVolumeService) CreateVolume(ctx context.Context, node, dc, name string, requestGb int64) (string, error) {
-	logger.Info("k8s.CreateVolume called", "name", name, "node", node, "size_gb", requestGb)
+func (s *LogicalVolumeService) CreateVolume(ctx context.Context, node, dc, name, parentID string, requestGb int64) (string, error) {
+	logger.Info("k8s.CreateVolume called", "name", name, "node", node, "size_gb", requestGb, "parentID", parentID)
 	s.mu.Lock()
 	defer s.mu.Unlock()
+	var lv *topolvmv1.LogicalVolume
+	// if the create volume request has no parent, proceed with regular lv creation.
+	if parentID == "" {
+		lv = &topolvmv1.LogicalVolume{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "LogicalVolume",
+				APIVersion: "topolvm.cybozu.com/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: topolvmv1.LogicalVolumeSpec{
+				Name:        name,
+				NodeName:    node,
+				DeviceClass: dc,
+				Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
+			},
+		}
 
-	lv := &topolvmv1.LogicalVolume{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "LogicalVolume",
-			APIVersion: "topolvm.cybozu.com/v1",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: name,
-		},
-		Spec: topolvmv1.LogicalVolumeSpec{
-			Name:        name,
-			NodeName:    node,
-			DeviceClass: dc,
-			Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
-		},
+	} else {
+		// On the other hand, if a volume has a datasource, create a thin snapshot of the parent volume with READ-WRITE access.
+		lv = &topolvmv1.LogicalVolume{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "LogicalVolume",
+				APIVersion: "topolvm.cybozu.com/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: topolvmv1.LogicalVolumeSpec{
+				Name:        name,
+				NodeName:    node,
+				DeviceClass: dc,
+				Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
+				Type:        "thin",
+				SourceID:    parentID,
+				AccessType:  "rw",
```

Comment:
- just confirming, for both clones and restore ops the access type should be `rw` and this doesn't have any relevance which creating snaps isn't it?

Reply:
Yes, for both restores and clones, the accessType should be `rw`.
It is relevance when we activate the snapshots as for `rw` we activate the snapshots for Read/Write access; on the other hand just `ro` for Read-Only access(In case of k8s Volumesnapshots)